src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs

```diff
@@ -120,9 +120,98 @@ internal class SemanticTokensHelpers
+                {
+                    // This is already a single-line span, so no modification is necessary.
+                    updatedClassifiedSpans.Add(span);
+                }
+            }
+
+            return updatedClassifiedSpans.ToArray();
+
+            static int ConvertToSingleLineSpan(
+                SourceText text,
+                ClassifiedSpan[] originalClassifiedSpans,
+                ArrayBuilder<ClassifiedSpan> updatedClassifiedSpans,
+                int spanIndex,
+                string classificationType,
+                int startLine,
+                int startOffset,
+                int endLine,
+                int endOffSet)
+            {
+                var numLinesInSpan = endLine - startLine + 1;
+                Contract.ThrowIfTrue(numLinesInSpan < 1);
+
+                var updatedSpanIndex = spanIndex;
+
+                for (var currentLine = 0; currentLine < numLinesInSpan; currentLine++)
+                {
+                    TextSpan? textSpan;
+
+                    // Case 1: First line of span
+                    if (currentLine == 0)
+                    {
+                        var absoluteStartOffset = text.Lines[startLine].Start + startOffset;
+                        var spanLength = text.Lines[startLine].End - absoluteStartOffset;
+                        textSpan = new TextSpan(absoluteStartOffset, spanLength);
+                    }
+                    // Case 2: Any of the span's middle lines
+                    else if (currentLine != numLinesInSpan - 1)
+                    {
+                        textSpan = text.Lines[startLine + currentLine].Span;
+                    }
+                    // Case 3: Last line of span
+                    else
+                    {
+                        textSpan = new TextSpan(text.Lines[endLine].Start, endOffSet);
+                    }
+
+                    var updatedClassifiedSpan = new ClassifiedSpan(textSpan.Value, classificationType);
+                    updatedClassifiedSpans.Add(updatedClassifiedSpan);
+
+                    // Since spans are expected to be ordered, when breaking up a multi-line span, we may have to insert
+                    // other spans in-between. For example, we may encounter this case when breaking up a multi-line verbatim
+                    // string literal containing escape characters:
+                    //     var x = @"one ""
+                    //               two";
+                    // The check below ensures we correctly return the spans in the correct order, i.e. 'one', '""', 'two'.
+                    while (updatedSpanIndex + 1 < originalClassifiedSpans.Length &&
+                        textSpan.Value.Contains(originalClassifiedSpans[updatedSpanIndex + 1].TextSpan))
+                    {
+                        updatedClassifiedSpans.Add(originalClassifiedSpans[updatedSpanIndex + 1]);
+                        updatedSpanIndex++;
+                    }
```

Comment:
i don't really understand what's going on here.  why do we need to do this?

Reply:
When we process the spans and turn them into tokens, we expect the spans to be ordered by start index. If we don't include this check, then we could end up with spans that are incorrectly ordered. This could in turn render the tokens we return to LSP inaccurate.

For instance, with the following code:
```C#
var x = @"one ""
two";
```

Without the check we would process the spans in the following order: `one`, `two`, `""`.
But with the check, we will correctly process them in the correct order: in `one`, `""`, `two`.

Reply:
I've also added a test to ensure we're doing the right thing in the case above.